packages/client/src/features/tools/node-creation/container-manager.ts

```diff
@@ -44,21 +44,36 @@ export interface TrackedInsert {
     options: InsertOptions;
 }
 
+export interface IContainerManager {
+    insert(proxy: GModelElement, location: Point, elementTypeId: string, opts?: Partial<InsertOptions>): TrackedInsert;
+    findContainer(location: Point, ctx: GModelElement, evt?: MouseEvent): ContainerElement | undefined;
+    addInsertFeedback(feedback: FeedbackEmitter, trackedInsert: TrackedInsert, ctx?: GModelElement, event?: MouseEvent): FeedbackEmitter;
+}
+
+/**
+ * The default {@link IContainerManager} implementation.
+ * This class class manages the insertion of elements into containers by validating their positions and types,
+ * providing feedback on the insertion process, and determining the appropriate container based on the location and context.
+ */
 @injectable()
-export class ContainerManager {
+export class ContainerManager implements IContainerManager {
     @inject(ChangeBoundsManager) protected readonly changeBoundsManager: ChangeBoundsManager;
 
     insert(proxy: GModelElement, location: Point, elementTypeId: string, opts?: Partial<InsertOptions>): TrackedInsert {
         const options = { ...DEFAULT_INSERT_OPTIONS, ...opts };
         const container = this.findContainer(location, proxy, opts?.evt);
-        let valid = !container || container.isContainableElement(elementTypeId);
+        let valid = this.isCreationAllowed(container, elementTypeId, opts);
         if (valid && (!container || options.validateLocationInContainer)) {
             // we need to check whether the location is valid either because we do not have a container or the option is set
             valid = opts?.validLocationOverwrite ?? this.changeBoundsManager.hasValidPosition(proxy, location);
         }
         return { elementTypeId, container, location, valid, options };
     }
 
+    protected isCreationAllowed(container: ContainerElement | undefined, elementTypeId: string, opts?: Partial<InsertOptions>): boolean {
```

Comment:
This is the core change of this PR, right? By default we delegate here to a method that is provided automatically by the `ApplyTypeHintsCommand` which considers the `containableElementTypeIds` set by the server.

I see how a customization might make sense given that we have access to the insert options here but I wonder in general whether it wouldn't be more useful to simply extend the `isContainableElement` method with those arguments and potentially even allow subclasses (e.g., the CategoryNode) to provide a custom implementation of it and only apply the default implementation from the `ApplyTypeHintsCommand` if no custom implementation is given.

What do you think?

Reply:
Yes, that's correct, this is the main change to help our use case. In our case, we use a more complex logic that relies on another compatibility service rather than just element types. Since we can't inject this service directly into the model element `isContainableElement` method, we needed to customize it up to this point.

Reply:
I see, so that is fine by me then. We should just make sure that we use that service everywhere then, e.g., I see a similar usage in `KeyboardPointerPosition` maybe that class should also use the service? Could mean that we need to make `isCreationAllowed` public but I'll leave that up to you to see if that is a good fit.

Reply:
Thank you Martin for spotting this! Yes, I agree, that would make sense then to re-use this here.
I'll adapt the code accordingly.
However, the accessibility node creation seems to be broken (also on master). So created a follow-up for this: https://github.com/eclipse-glsp/glsp/issues/1394

Reply:
Thank you for creating the issue.